products/batch_exports/backend/temporal/pipeline/types.py

```diff
@@ -1,10 +1,17 @@
-import typing
+from dataclasses import dataclass
 
 
-class BatchExportResult(typing.NamedTuple):
+@dataclass
+class BatchExportResult:
     # This is the total number of records that were successfully exported
     # (not the number of record batches, but the number of records in all record batches)
-    records_completed: int
+    records_completed: int | None = None
     # This is the number of bytes of data exported (i.e. not the number of bytes in ClickHouse or the internal stage)
     # and therefore takes into account things like the file type and compression
-    bytes_exported: int
+    bytes_exported: int | None = None
+    # This is the error that occurred, if any
+    error: str | None = None
+
+    @classmethod
+    def from_exception(cls, e: Exception) -> "BatchExportResult":
+        return cls(error=f"{e.__class__.__name__}: {e}")
```

Comment:
Hmm, maybe separate this into `error_type` and `error_message`? Just to have the flexibility when formatting logs for the user.

Reply:
I guess our current logging is just doing the same thing, so this could be a bit too much of future-proofing. But I rather  be flexible giving that it can be annoying to update input/output types with temporal after defined.